src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs

```diff
@@ -53,19 +55,58 @@ public sealed class AnalyzerFileReference : AnalyzerReference, IEquatable<Analyz
         /// <param name="assemblyLoader">Loader for obtaining the <see cref="Assembly"/> from the <paramref name="fullPath"/></param>
         public AnalyzerFileReference(string fullPath, IAnalyzerAssemblyLoader assemblyLoader)
         {
-            _fullPath = fullPath ?? throw new ArgumentNullException(nameof(fullPath));
+            CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
+
+            FullPath = fullPath;
+            _assemblyLoader = assemblyLoader ?? throw new ArgumentNullException(nameof(assemblyLoader));
+
             _diagnosticAnalyzers = new Extensions<DiagnosticAnalyzer>(this, IsDiagnosticAnalyzerAttribute, GetDiagnosticsAnalyzerSupportedLanguages);
             _generators = new Extensions<ISourceGenerator>(this, IsGeneratorAttribute, GetGeneratorsSupportedLanguages);
-            _assemblyLoader = assemblyLoader ?? throw new ArgumentNullException(nameof(assemblyLoader));
 
             // Note this analyzer full path as a dependency location, so that the analyzer loader
             // can correctly load analyzer dependencies.
-            if (PathUtilities.IsAbsolute(fullPath))
+            assemblyLoader.AddDependencyLocation(fullPath);
+        }
+
+        public override bool Equals(object? obj)
+            => Equals(obj as AnalyzerFileReference);
+
+        public bool Equals(AnalyzerFileReference? other)
+        {
+            if (ReferenceEquals(this, other))
             {
-                assemblyLoader.AddDependencyLocation(fullPath);
+                return true;
             }
+
+            return other is object &&
+                ReferenceEquals(_assemblyLoader, other._assemblyLoader) &&
+                FullPath == other.FullPath;
         }
 
+        // legacy, for backwards compat:
+        public bool Equals(AnalyzerReference? other)
+        {
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
+
+            if (other is null)
+            {
+                return false;
+            }
+
+            if (other is AnalyzerFileReference fileReference)
+            {
+                return Equals(fileReference);
+            }
+
+            return FullPath == other.FullPath;
```

Comment:
>FullPath == other.FullPath [](start = 19, length = 26)

When is this `true` for distinct `AnalyzerReference` types?

Reply:
This is for backward compat. We shouldn't really be comparing different type of references, but we did implement it this way so need to keep it.
Note that anyone can subclass AnalyzerReference and then compare AnalyzerFileReference to it.